squid: libcephfs client cache trimming might remove dentries attached to opened dir inode.

dparmar18 opened the pull request:
Backport of https://github.com/ceph/ceph/pull/60909
Backport tracker: https://tracker.ceph.com/issues/69906
Parent tracker: https://tracker.ceph.com/issues/69092


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

ifed01 committed:
test/libcephfs: reproduce an issue after client cache trimming

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>
(cherry picked from commit a27ed27c2c2667aad171d66574e446f7121fdc36)

ifed01 committed:
libcephfs/client: pin inode/dentry for an opened directory

Fixes: https://tracker.ceph.com/issues/69092

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>
(cherry picked from commit 7b8a86f105f66eb0d39b6c4920a722eedd88fedd)

ifed01 committed:
test/libcephfs: remove warning in Windows build

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>
(cherry picked from commit c9d73451e74f76640336d27ddcfbaa96964c706f)

ifed01 committed:
common/ceph_fs: Enable O_DIRECTORY|O_NOFOLLOWUP flags translation under Win
in ceph_flags_sys2wire()

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>
(cherry picked from commit 4badc83940fd7c072fd28fe81892c98304d03502)

dparmar18 commented:
@ifed01 please re-review

dparmar18 commented:
> @dparmar18 Please resolve the conflict

done

ifed01 reviewed (changes requested):
make check failure looks relevant

dparmar18 commented:
> make check failure looks relevant

```
FAILED: src/client/CMakeFiles/client.dir/Client.cc.o 
/usr/bin/ccache /usr/bin/clang++-14 -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_HAS_IO_URING -DBOOST_ASIO_NO_TS_EXECUTORS -DBOOST_CHRONO_DYN_LINK -DBOOST_CHRONO_NO_LIB -DBOOST_DATE_TIME_DYN_LINK -DBOOST_DATE_TIME_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_IOSTREAMS_NO_LIB -DBOOST_LOCALE_DYN_LINK -DBOOST_LOCALE_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_RANDOM_DYN_LINK -DBOOST_RANDOM_NO_LIB -DBOOST_SYSTEM_DYN_LINK -DBOOST_SYSTEM_NO_LIB -DBOOST_THREAD_DYN_LINK -DBOOST_THREAD_NO_LIB -DHAVE_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/home/jenkins-build/build/workspace/ceph-pull-requests/build/src/include -I/home/jenkins-build/build/workspace/ceph-pull-requests/src -isystem /opt/ceph/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/build/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/api/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/exporters/jaeger/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/ext/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/sdk/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/xxHash -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/fmt/include -g -Werror -fPIC -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -DBOOST_PHOENIX_STL_TUPLE_H_ -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -Wno-address-of-packed-member -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register -DCEPH_DEBUG_MUTEX -D_GLIBCXX_ASSERTIONS -fdiagnostics-color=auto -std=c++20 -MD -MT src/client/CMakeFiles/client.dir/Client.cc.o -MF src/client/CMakeFiles/client.dir/Client.cc.o.d -o src/client/CMakeFiles/client.dir/Client.cc.o -c /home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:10435:25: error: no viable conversion from 'const InodeRef' (aka 'const intrusive_ptr<Inode>') to 'Inode *'
      *fhp = _create_fh(in, flags, cmode, perms);
                        ^~
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:10270:31: note: passing argument to parameter 'in' here
Fh *Client::_create_fh(Inode *in, int flags, int cmode, const UserPerm& perms)
                              ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:10442:17: error: no viable conversion from 'const InodeRef' (aka 'const intrusive_ptr<Inode>') to 'Inode *'
        _ll_get(in);
                ^~
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.h:1659:23: note: passing argument to parameter 'in' here
  void _ll_get(Inode *in);
                      ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:10500:13: error: use of undeclared identifier 'CEPHFS_EBADF'
    return -CEPHFS_EBADF;
            ^
3 errors generated.
```

joscollin commented:
jenkins test make check

tchaikov committed:
test/libcephfs: reduce SnapDiffDeletionRecreation bulk_count on Windows

this test timed out on Windows. and HugeSnapDiffLargeDelta, at half
the file count, passed in 508 seconds on the same run, suggesting this
test takes ~17 minutes on Windows -- beyond the test runner limit.

we haven't profiled the Windows client yet, but the likely culprit is
EventPoll, the Windows messenger backend, which scans the entire poll
array on every event_wait() and poll_ctl() call rather than using a
keyed data structure.

in this change, we reduce bulk_count to 1 << 12 on Windows. the unique
thing this test covers is the deletion-recreation pattern: a name that
exists as a file in snap1, gets deleted, and reappears as a directory in
snap2 -- it must show up in the diff with both snapids. 4096 produces
1024 such pairs, which is enough to exercise that logic. multi-fragment
snapdiff is already covered by HugeSnapDiffLargeDelta, which derives its
file count from mds_bal_split_size and mds_bal_fragment_fast_factor
explicitly to trigger fragmentation.

Fixes: https://tracker.ceph.com/issues/77015
Signed-off-by: Kefu Chai <k.chai@proxmox.com>
(cherry picked from commit 4369bea1475e82523d6430e055b6a91f24a40ee0)

dparmar18 committed:
client: fix compilation errors

1) FAILED: src/client/CMakeFiles/client.dir/Client.cc.o
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:10435:25: error: no viable conversion from 'const InodeRef' (aka 'const intrusive_ptr<Inode>') to 'Inode *'
      *fhp = _create_fh(in, flags, cmode, perms);
                        ^~
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:10270:31: note: passing argument to parameter 'in' here
Fh *Client::_create_fh(Inode *in, int flags, int cmode, const UserPerm& perms)
                              ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:10442:17: error: no viable conversion from 'const InodeRef' (aka 'const intrusive_ptr<Inode>') to 'Inode *'
        _ll_get(in);
                ^~
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.h:1659:23: note: passing argument to parameter 'in' here
  void _ll_get(Inode *in);

Fix from https://github.com/ceph/ceph/commit/b8662fdb03373ed6d29c47efa94cd391a7c9e740

2) /home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:10500:13: error: use of undeclared identifier 'CEPHFS_EBADF'
    return -CEPHFS_EBADF;

Fix from https://github.com/ceph/ceph/commit/7055fa53c50062c285386db488bcdaee57c5dd1c

Signed-off-by: Dhairya Parmar <dparmar@redhat.com>

dparmar18 commented:
/audit retest

dparmar18 commented:
this should be okay now. I've fixed the compilation errors in the last commit.

dparmar18 commented:
@vshankar you'd need to override releng audit, it won't pass because of the last commit (since it doesn't exist in parent PR)

dparmar18 commented:
windows test failed with:
```
[2026-06-02T13:32:08.000Z] [isolated][googletest] ceph_test_libcephfs_snapdiff failed. Error: Traceback (most recent call last):
[2026-06-02T13:32:08.000Z] [isolated][googletest] ceph_test_libcephfs_suidsgid passed.
[2026-06-02T13:32:08.000Z] [isolated][googletest] ceph_test_libcephfs_vxattr passed.
[2026-06-02T13:32:08.000Z] [isolated][googletest] unittest_admin_socket passed.
ceph_test_libcephfs_snapdiff
The process tried to write to a nonexistent pipe.
[2026-06-02T13:32:11.000Z] subunit-trace reports failures: Command failed: type C:\workspace\test_results\subunit.out | subunit-trace
One or more test suites have failed
At C:\workspace\repos\ceph-win32-tests\test_host\run_tests.ps1:28 char:9
+         throw "One or more test suites have failed"
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (One or more test suites have failed:String) [], RuntimeException
    + FullyQualifiedErrorId : One or more test suites have failed
```
https://jenkins.ceph.com/job/ceph-windows-pull-requests/79925/consoleText

vshankar commented:
> windows test failed with:

This seems related. Have you checked?

dparmar18 commented:
> > windows test failed with:
> 
> This seems related. Have you checked?

there is nothing but this in the log,

vshankar commented:
> there is nothing but this in the log,

There a couple of ways to go about this. First check if there are some windows related commits that need included. This change has one, but there could be more. Secondly, if we run out of options, the we can seek help from @petrutlucian94

dparmar18 commented:
jenkins test windows

dparmar18 commented:
> > there is nothing but this in the log,
> 
> There a couple of ways to go about this. First check if there are some windows related commits that need included. This change has one, but there could be more. Secondly, if we run out of options, the we can seek help from @petrutlucian94

those changes were also present in the PR targeted to main branch and that passed. I also checked https://github.com/ceph/ceph-win32-tests/tree/main and it doesnt reflect much changes, only plausible rationale i have is the snapdiff test cases hit timeout https://github.com/ceph/ceph-win32-tests/blob/main/test_host/run_tests.ps1#L5:
```
[2026-06-02T13:01:01.000Z] Finished 4 out of 8 jobs.
...
[2026-06-02T13:30:48.000Z] Finished 4 out of 8 jobs.
```
IMO this is an intermittent issue and is hitting due to heavy tests that create large directory trees and many snapshots like `HugeSnapDiffSmallDelta`, `HugeSnapDiffLargeDelta` and `SnapDiffDeletionRecreation`.

I've re-triggered the windows build and there is nothing more i can do without knowing if it passed/failed. It's close to midnight here so i'll have to call it a day and i'll try to get it done before tomorrow EOD. If this needs to be fixed right now then i think its better to be hand over to someone in their working hours.

dparmar18 commented:
i think we need https://github.com/ceph/ceph/pull/69203

dparmar18 commented:
jenkins test docs

dparmar18 commented:
```
The following tests FAILED:
	238 - unittest_posix_bucket_cache (Failed)
```

dparmar18 commented:
jenkins test make check

dparmar18 commented:
good to merge post overriding releng audit check

vshankar reviewed (approved):
Well done @dparmar18

vshankar commented:
> good to merge post overriding releng audit check

cc @joscollin for backport QA.

joscollin commented:
This PR is under test in [https://tracker.ceph.com/issues/77133](https://tracker.ceph.com/issues/77133).

joscollin reviewed (approved):
No src/test failures found.
